src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -370,126 +418,410 @@ private sealed class SendFileOperation : WriteOperation
+#endif
+            }
+
+            public void Dispose()
+            {
+                Debug.Assert(Monitor.IsEntered(_lockObject));
+                Monitor.Exit(_lockObject);
+            }
         }
 
         private struct OperationQueue<TOperation>
             where TOperation : AsyncOperation
         {
+            // Quick overview:
+            // 
+            // When attempting to perform an IO operation, the caller first checks IsReady,
+            // and if true, attempts to perform the operation itself.
+            // If this returns EWOULDBLOCK, or if the queue was not ready, then the operation
+            // is enqueued by calling StartAsyncOperation and the state becomes Waiting.
+            // When an epoll notification is received, we check if the state is Waiting,
+            // and if so, change the state to Processing and enqueue a workitem to the threadpool 
+            // to try to perform the enqueued operations.
+            // If an operation is successfully performed, we remove it from the queue,
+            // enqueue another threadpool workitem to process the next item in the queue (if any),
+            // and call the user's completion callback.
+            // If we successfully process all enqueued operations, then the state becomes Ready;
+            // otherwise, the state becomes Waiting and we wait for another epoll notification.
+
+            private enum QueueState
+            {
+                Ready = 0,          // Indicates that data MAY be available on the socket.
+                                    // Queue must be empty.
+                Waiting = 1,        // Indicates that data is definitely not available on the socket.
+                                    // Queue must not be empty.
+                Processing = 2,     // Indicates that a thread pool item has been scheduled (and may 
+                                    // be executing) to process the IO operations in the queue.
+                                    // Queue must not be empty.
+                Stopped = 3,        // Indicates that the queue has been stopped because the 
+                                    // socket has been closed.
+                                    // Queue must be empty.
+            }
+
+            // These fields define the queue state.
+
+            private QueueState _state;      // See above
+            private int _sequenceNumber;    // This sequence number is updated when we receive an epoll notification.
+                                            // It allows us to detect when a new epoll notification has arrived
+                                            // since the last time we checked the state of the queue.
+                                            // If this happens, we MUST retry the operation, otherwise we risk
+                                            // "losing" the notification and causing the operation to pend indefinitely.
```

Comment:
Presumably we're not concerned about 4B notifications causing this to loop around and cause confusion?